tests: Delete VMI prior to NFS server pod

vasiliy-ul opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

Kubelet fails to umount the NFS volume and gets stuck when performing pod cleanup if the server has already gone. Ensure that the VMI is deleted first, and the corresponding volume is released (current global tests cleanup hook does not guarantee deletion order).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

I observe the issue after running nightly tests. I see the nodes in `NotReady` state with stale NFS mount points:

```
# mount | grep nfs
10.42.1.220:/ on /var/lib/kubelet/pods/0747bcc2-9f8d-473d-acfc-21393679e9d0/volumes/kubernetes.io~nfs/test-nfsfb92r6g7cvw57rghvh25q7j99wj7vws9tbbqwhkp58v222mq type nfs4 (rw,relatime,vers=4.2,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=192.168.122.132,local_lock=none,addr=10.42.1.220)
10.42.1.225:/ on /var/lib/kubelet/pods/16b805f0-0e86-469d-8bf8-7d8685830a3e/volumes/kubernetes.io~nfs/test-nfsfb92r6g7cvw57rghvh25q7j99wj7vws9tbbqwhkp58v222mq type nfs4 (rw,relatime,vers=4.2,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=192.168.122.132,local_lock=none,addr=10.42.1.225)
10.42.1.240:/ on /var/lib/kubelet/pods/d6ccc012-5529-438b-b3fc-f3d35df04ab1/volumes/kubernetes.io~nfs/test-nfsfb92r6g7cvw57rghvh25q7j99wj7vws9tbbqwhkp58v222mq type nfs4 (rw,relatime,vers=4.2,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=192.168.122.132,local_lock=none,addr=10.42.1.240)
```

The nodes become non-operational and the subsequent tests fail. This can only be fixed by either manually running `umount -lf ...` or by rebooting the affected nodes.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

vasiliy-ul commented:
/retest-required

vasiliy-ul commented:
/retest-required

vasiliy-ul commented:
/retest-required

dhiller commented:
/test pull-kubevirt-check-tests-for-flakes

dhiller reviewed (commented):
@vasiliy-ul the changes look good to me, thank you for your work!

However, it seems that `pull-kubevirt-check-tests-for-flakes` sometimes stumbles over the exact test set that the `AfterSuite` is applied to.

Please note that the test execution order in the lane mentioned above is randomized. Also the test is non serial, so there might be two options:
* an order dependence is lurking inside the tests or
* maybe the non serial tests are not properly deserialized, i.e. not using the test namespaces.

ICYMI for some sparse docs see [here](https://github.com/kubevirt/kubevirt/blob/8ad9647be865c0e8648dc4379c7c79b828408fbf/tests/README.md?plain=1#L3)

vasiliy-ul commented:
Hi @dhiller. Thank you for bringing this up. I looked at the issue and have a few thoughts and observations to share:
- Firstly, unfortunately, I was not able to reproduce that exact failure locally. It seems to heavily depend on random factors.
- When looking at the tests code, I couldn't spot any obvious bugs which could lead to that issue. The only thing that I’ve found is a leftover resource in one of the tests. But that is unlikely to cause troubles since after each run the CI is nuked. Anyway, I added a fix for that in a separate commit. PTAL.
- I found it a bit weird that two of the NFS tests are marked as `Serial`:
  https://github.com/kubevirt/kubevirt/blob/bb2f77e8248b6adbc7723baee64449966b9fb0d6/tests/storage/storage.go#L288-L290
  https://github.com/kubevirt/kubevirt/blob/bb2f77e8248b6adbc7723baee64449966b9fb0d6/tests/storage/storage.go#L443-L444
  From my observation, all the NFS tests from that suite can be run in parallel as they do not alter the environment.
- While looking at the artifacts from the flaky test failure https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/10310/pull-kubevirt-check-tests-for-flakes/1693871238436360192, I noticed in the virt-launcher's logs a delay of 90 seconds before libvirt starts qemu. The same delay can be observed when running any of the NFS tests locally, and it does not depend on the amount of nodes, pods, network latency, etc. It stays constant. Due to this delay, we can also see in the test logs `unknown error encountered sending command SyncVMI: rpc error: code = DeadlineExceeded desc = context deadline exceeded` warnings from virt-handler. After diving a bit deeper, I've found that the delay is caused by the NFS option, which defines a grace period after server start/restart. During that period, the clients will stall if they try to access the NFS share. Currently, this setting is hardcoded in the image that is used as a base for the nfs-server:  https://github.com/slintes/docker-nfs-ganesha/blob/6300c3f9a522d0f11f6b26bb8743bc8340c9303b/rootfs/opt/start_nfs.sh#L40. I think this delay of 90 seconds may cause the flaky behavior. In fact, I am not sure that this setting brings any benefit when used in the CI. So I added a commit to disable it completely. The tests are now executed much faster. The implementation is a bit hacky, but if we are okay with the approach in general, I can submit a patch to the base image to make it more flexible and configurable. Alternatively, if we still want to keep the setting, then we need to consider the grace period in the timeouts used for the tests (which is IMHO not nice, since it will be just a big time waste).

vasiliy-ul commented:
For reference: https://github.com/nfs-ganesha/nfs-ganesha/blob/next/src/doc/man/ganesha-core-config.rst#nfsv4-

vasiliy-ul commented:
Looking at the storage lane failure: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/10310/pull-kubevirt-e2e-k8s-1.27-sig-storage/1694671859703877632
```
                "message": "Pod was rejected: The node had condition: [DiskPressure]. ",
                "reason": "Evicted",
```
Seems that we may need more storage for the test jobs...

vasiliy-ul commented:
> @vasiliy-ul might that also be the reason why those two nfs tests run in serial mode?

I do not know the reasoning behind making those tests Serial, unfortunately. Found this old issue https://github.com/kubevirt/kubevirt/issues/4962 which seems related.

awels commented:
So from what I know when a node gets disk pressure it is because the OS is running out of space. Which is likely to happen because are using the local volume provisioner on the same partition as the OS, we just create some bind mounts in a directory somewhere. In order for this to get fix we just need to increase the size of the VM disk in kubevirtci. I honestly don't remember the last time I made a chance to kubevirtci. But I suspect it is pretty straight forward to change the disk size of the VM.

awels commented:
> btw, I do not think that `[DiskPressure]` is related to the `local-storage-provider` and volume provisioning. This condition usually means that the free storage space on the node is low and kubelet cannot schedule pods. But maybe I am missing smth...

You are right normally, but due to the bind mounts the local volume provisioner gives access to the same filesystem as the OS and it is entirely possible to exhaust the space available to the OS/Kubelet that way.

dhiller commented:
> So from what I know when a node gets disk pressure it is because the OS is running out of space. Which is likely to happen because are using the local volume provisioner on the same partition as the OS, we just create some bind mounts in a directory somewhere. In order for this to get fix we just need to increase the size of the VM disk in kubevirtci. I honestly don't remember the last time I made a chance to kubevirtci. But I suspect it is pretty straight forward to change the disk size of the VM.

Is [this](https://github.com/kubevirt/kubevirtci/blob/f661bfe0e3678e5409c057855951c50a912571a0/cluster-provision/centos9/scripts/vm.sh#L96) it?

dhiller commented:
^^ @awels

awels commented:
Yes I think so, the current VM disk is 35G, so I would simply increase that, it is a sparse file so it shouldn't impact the size of any containers.

dhiller commented on images/nfs-server/entrypoint.sh:
```diff
@@ -0,0 +1,9 @@
+set -euxo pipefail
+
+sed -i"" \
+    -e "s#Grace_Period = 90#Graceless = true#g" \
```
nit: an explanation why we are doing this would be nice here

dhiller reviewed (approved):
@vasiliy-ul I don't want to block on investigating further the check-test-for-flakes, I think a later run showed no failure, so that is good enough for me.

Other than that I only found one nit, you decide whether to adress it, in the latter case just remove the hold.

/approve
/hold

dhiller reviewed (approved):
@vasiliy-ul thank you for your work! :pray:

vasiliy-ul commented:
/unhold

vasiliy-ul commented:
/cherry-pick release-1.0

vasiliy-ul commented:
Is the CI stuck here? The tests are not triggered and has been in `Expected` state for a while. @dhiller, any thought?

vasiliy-ul commented:
> Is the CI stuck here? The tests are not triggered and have been in `Expected` state for a while. @dhiller, any thought?

Ah, seems that a force-push after amend/no-edit helped.

vasiliy-ul commented:
Hi @dhiller, could you re-lgtm the PR? I had to force-push to trigger the stuck CI here and hence lost the approval. No source code changes made.

dhiller reviewed (approved):
@vasiliy-ul sorry for the delay :-/

brianmcarey commented:
/retest-required
